feat(sdk): CLI - Allow upload_version by the pipeline's name. Fixes #3901

NikeNano opened the pull request:
**Description of your changes:**
This pr updates the KFP cli to allow pipeline upload_version to accept name instead of id. This will make is easier for some users and allow for a larger flexibility. 

This should fix PR: https://github.com/kubeflow/pipelines/issues/3901

We could also the check that only one/and always one of pipeline_name and pipeline_is using click, whould this be prefered?

Bobgy commented on sdk/python/kfp/_client.py:
```diff
@@ -304,6 +304,25 @@ def create_experiment(self, name, description=None, namespace=None):
     return experiment
 
+  def get_pipeline_id(self, name, page_token='', page_size=100):
+    """Returns the pipeline id if a pipeline with the name exsists.
```
Note, when multi user mode enabled, you can only query pipelines by namespace.

NikeNano replied:
Could you elaborate further on this @Bobgy? Do you mean I should also pass in the namespace as a filter? Or do you mean that the results will be limited to the namespace I have access to? I have checked the API and they don't mention the use of namespace as a parameter: https://www.kubeflow.org/docs/pipelines/reference/api/kubeflow-pipeline-api-spec/#tag-PipelineService

Bobgy commented on sdk/python/kfp/_client.py:
```diff
@@ -304,6 +304,25 @@ def create_experiment(self, name, description=None, namespace=None):
+      A response object including a list of experiments and next page token.
+    """
+    while True:
+        result = self.list_pipelines(page_size=page_size, page_token=page_token)
```
this doesn't scale when number of pipelines is huge.
list_pipelines supports passing a filter, UI client's usage can be found at https://github.com/kubeflow/pipelines/blob/master/frontend/src/pages/GettingStarted.tsx#L126.

You may try to learn this usage.

NikeNano commented:
Thanks for the feedback @Bobgy. I have not used protobuff before but gave it a try and got it working at least  :)! I am not sure where it would make the most sense to put `filter_pb2.py` and if it should be auto generated and if so how? Concerning the filtering on namespace I am not sure how we can do it? Is it allowed to do the same way?

Ark-kun commented on sdk/python/kfp/cli/pipeline.py:
```diff
@@ -57,10 +63,18 @@ def upload(ctx, pipeline_name, package_file):
+    if pipeline_name!=None: 
+        pipeline_id = client.get_pipeline_id(name=pipeline_name)
+        if pipeline_id==None: 
+            logging.warning(f"Cant find a pipeline with name: {pipeline_name}")
```
f-strings are not supported in python 3.5

Ark-kun commented on sdk/python/kfp/cli/pipeline.py:
```diff
@@ -57,10 +63,18 @@ def upload(ctx, pipeline_name, package_file):
     """Upload a version of the KFP pipeline"""
     client = ctx.obj["client"]
-
+    if pipeline_id==None and pipeline_name==None: #[TODO] do this using click instead... 
```
Style: There need to be spaces around the comparison operators.

Ark-kun commented on sdk/python/kfp/cli/pipeline.py:
```diff
@@ -57,10 +63,18 @@ def upload(ctx, pipeline_name, package_file):
     client = ctx.obj["client"]
-
+    if pipeline_id==None and pipeline_name==None: #[TODO] do this using click instead... 
+        logging.Warning("Need to suppy 'pipeline-name' or 'pipeline-id'")
```
This probably needs to be an error.

NikeNano replied:
Will update with value error on both places.

Ark-kun commented on sdk/python/kfp/cli/pipeline.py:
```diff
@@ -57,10 +63,18 @@ def upload(ctx, pipeline_name, package_file):
+    if pipeline_name!=None: 
+        pipeline_id = client.get_pipeline_id(name=pipeline_name)
+        if pipeline_id==None: 
+            logging.warning(f"Cant find a pipeline with name: {pipeline_name}")
```
This probably needs to be an error.

NikeNano replied:
Should we raise an `ValueError`?

NikeNano commented:
Thanks for the feedback @Ark-kun, resolved the comments.

NikeNano commented:
/test kubeflow-pipeline-sample-test

NikeNano commented on sdk/python/kfp/_client.py:
```diff
@@ -304,6 +307,32 @@ def create_experiment(self, name, description=None, namespace=None):
+      predicate.string_value=name 
+      filterName.predicates.append(predicate)
+      result = self._pipelines_api.list_pipelines(page_token=page_token, page_size=page_size, filter=urllib.parse.quote(MessageToJson(filterName)))
+      print(result)
```
```suggestion
```

NikeNano commented:
@Ark-kun I have updated but are getting some issues with ProtoBuf and it seems to be a issue related to versions for ProtoBuf. I have updated the packages in the requirement file but are builds/test using dependencies from somewhere else? 

This threads talks about the issue: https://stackoverflow.com/questions/61922334/how-to-solve-attributeerror-module-google-protobuf-descriptor-has-no-attribu

Bobgy commented on sdk/python/kfp/filter_pb2.py:
```diff
@@ -0,0 +1,409 @@
+# -*- coding: utf-8 -*-
+# Generated by the protocol buffer compiler.  DO NOT EDIT!
```
How did you generate this file? Would you mind adding related commands in https://github.com/kubeflow/pipelines/blob/master/backend/api/build_kfp_server_api_python_package.sh? and let this be part of the `kfp-server-api` package, because is this generated from backend api.

Bobgy commented:
updated title to better comply with our convention

NikeNano commented on sdk/python/kfp/_client.py:
```diff
@@ -304,6 +307,31 @@ def create_experiment(self, name, description=None, namespace=None):
+      filterName = filter_pb2.Filter()
+      predicate = filter_pb2.Predicate() 
+      predicate.key = "name"
+      predicate.op = 1
```
Created a dict which maps the operation type to the correct key.

NikeNano commented:
Thanks for the feedback @Bobgy. I still have some issues with the  I have updated but are getting some issues with ProtoBuf and it seems to be an issue related to versions for ProtoBuf and Python 3.5. 

This threads talks about the issue: https://stackoverflow.com/questions/61922334/how-to-solve-attributeerror-module-google-protobuf-descriptor-has-no-attribu

I tried to use different versions of Protobuf but don't get it right ... do you know more about it?

Bobgy commented:
> Thanks for the feedback @Bobgy. I still have some issues with ProtoBuf and it seems to be an issue related to versions for ProtoBuf and Python 3.5.
> 
> This threads talks about the issue: https://stackoverflow.com/questions/61922334/how-to-solve-attributeerror-module-google-protobuf-descriptor-has-no-attribu
> 
> I tried to use different versions of Protobuf but don't get it right ... do you know more about it?

Good luck! I don't know more details about this.

/cc @numerology @Ark-kun @jingzhang36 
probably can help

Backup plan: did you consider constructing the json payload by yourself, the filter package is just a helper, we may not need it at all like in frontend code sample:
https://github.com/kubeflow/pipelines/blob/f1d90407a8a2f56db11199c9c73e6df6c4a8b093/frontend/src/pages/GettingStarted.tsx#L169-L180

NikeNano commented:
> Backup plan: did you consider constructing the json payload by yourself, the filter package is just a helper, we may not need it at all like in frontend code sample. 

I though about it but, assumed the preferred way was to built the proper helper file. I will push an solution where we do it as the frontend :)

Bobgy commented on sdk/python/kfp/cli/pipeline.py:
```diff
@@ -57,10 +63,16 @@ def upload(ctx, pipeline_name, package_file):
-
+    if pipeline_id == None and pipeline_name == None: #[TODO] do this using click instead... 
+        raise ValueError("Need to suppy 'pipeline-name' or 'pipeline-id'")
+    if pipeline_name!=None: 
```
nit: shall we also raise a ValueError when both id and name are provided? (they may not be consistent, I think it's better we just avoid having both at the same time)

Bobgy commented on sdk/python/kfp/_client.py:
```diff
@@ -304,6 +312,31 @@ def create_experiment(self, name, description=None, namespace=None):
       IPython.display.display(IPython.display.HTML(html))
     return experiment
 
+  def get_pipeline_id(self, name, page_token='', page_size=100):
```
I think you can erase page_token and page_size here.

Bobgy commented on sdk/python/kfp/_client.py:
```diff
@@ -304,6 +312,31 @@ def create_experiment(self, name, description=None, namespace=None):
+    Args:
+      name: pipeline name
+      page_token: token for starting of the page.
+      page_size: size of the page.
```
reminder, also the docstring

NikeNano replied:
sorry, multi tasking while listening in to some meetings.

Bobgy commented:
> > Backup plan: did you consider constructing the json payload by yourself, the filter package is just a helper, we may not need it at all like in frontend code sample.
> 
> I though about it but, assumed the preferred way was to built the proper helper file. I will push an solution where we do it as the frontend :)

I agree proper helper is the preferred way, just that when it's hard to get it working. We can choose the practical path.

Bobgy commented:
Updated title to be more concise and indicates which issue it solved

Bobgy commented on sdk/python/kfp/_client.py:
```diff
@@ -304,6 +312,29 @@ def create_experiment(self, name, description=None, namespace=None):
       IPython.display.display(IPython.display.HTML(html))
     return experiment
 
+  def get_pipeline_id(self, name, page_size=100):
```
and did you forget the `page_size`?

Bobgy commented:
/lgtm
This is perfect!

/assign @Ark-kun @numerology 
Do you want to further review?

I think this is good to go.

numerology commented on sdk/python/kfp/cli/pipeline.py:
```diff
@@ -57,10 +63,16 @@ def upload(ctx, pipeline_name, package_file):
     """Upload a version of the KFP pipeline"""
     client = ctx.obj["client"]
-
+    if (pipeline_id == None and pipeline_name == None)  or (pipeline_id and pipeline_name): #[TODO] do this using click instead... 
```
nit: you can make it more succinct by

```
if bool(pipeline_id) == bool(pipeline_name):
  ...
```

NikeNano replied:
Cool, will update!

numerology commented on sdk/python/kfp/cli/pipeline.py:
```diff
@@ -57,10 +63,16 @@ def upload(ctx, pipeline_name, package_file):
+    if pipeline_name!=None: 
+        pipeline_id = client.get_pipeline_id(name=pipeline_name)
+        if pipeline_id==None: 
+            raise ValueError("Cant find a pipeline with name: {}".format(pipeline_name))
```
```suggestion
            raise ValueError("Can't find a pipeline with name: {}".format(pipeline_name))
```

numerology commented on sdk/python/kfp/cli/pipeline.py:
```diff
@@ -57,10 +63,16 @@ def upload(ctx, pipeline_name, package_file):
+        pipeline_id = client.get_pipeline_id(name=pipeline_name)
+        if pipeline_id==None: 
+            raise ValueError("Cant find a pipeline with name: {}".format(pipeline_name))
+    logging.info("The pipeline id is: {}".format(pipeline_id))
```
Consider using % formatter in logging when possible. This is b/c if we use `format`, the string template gets resolved even when not needed.

See explanation in
https://reinout.vanrees.org/weblog/2015/06/05/logging-formatting.html
for example.

NikeNano replied:
Didn't know, thanks :) I will update them!

NikeNano commented on sdk/python/kfp/_client.py:
```diff
@@ -39,7 +39,15 @@
+    # Operators on scalar values. Only applies to one of |int_value|,
+    # |long_value|, |string_value| or |timestamp_value|.
+    "EQUALS" : 1,
+    "NOT_EQUALS" : 2,
```
No, they are not I kept them on order to have a reference for the alternative since we don't have a proper helper file. Maybe we should remove it. Initially I only. set it to 1, but @Bobgy suggested that we added a mapping in order for readers of the code to understand what it meant.

Bobgy commented:
/lgtm
/approve

Thanks!

NikeNano commented:
/retest

Bobgy commented:
@googlebot rescan
